File: array/threesumclosest.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func threesum(nums []int, target int) int {
	sort.Ints(nums)
	var ans int
	closest := float64(1<<32 - 1)
	high := len(nums) - 1

	for k := 0; k < (high - 1); k++ {
		remain := target - nums[k]
		for i, j := k+1, high; i < j; {
			current_sum := nums[i] + nums[j]
			current_closest := math.Min(float64(closest), math.Abs(float64(remain)-float64(current_sum)))
			if current_closest < closest {
				closest = current_closest
				ans = nums[k] + current_sum
			}
			if remain == current_sum {
				return ans
			} else if remain > current_sum {
				i++
			} else {
				j--
			}
		}
	}

	return ans
}

func main() {
	nums := []int{-1, 2, 1, -4}
	result := threesum(nums, 1)
	fmt.Println(result)
}

File: array/threesumclosest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest_sum = float("inf")
        high = len(nums) - 1 
        ans = float("inf")
        for k in range(high-1):
            i = k+1
            j = high
            num = nums[k]
            remain = target - num
            while i < j:
                current_sum = nums[i] + nums[j]
                current_closest = min(closest_sum, abs(remain - current_sum))
                if current_closest < closest_sum:
                    closest_sum = current_closest
                    ans = current_sum + num
                if current_sum == remain:
                    return ans
                elif current_sum > remain:
                    j -= 1
                elif current_sum < remain:
                    i += 1
        return ans


if __name__ == "__main__":
    nums = [-1,2,1,-4]
    print(Solution().threeSumClosest(nums, -4))